packages/nifti-volume-loader/src/helpers/fetchAndAllocateNiftiVolume.ts

```diff
@@ -0,0 +1,266 @@
+};
+
+export default async function fetchAndAllocateNiftiVolume(
+  volumeId: string
+): Promise<NiftiImageVolume> {
+  // nifti volumeIds start with 'nifti:' so we need to remove that
+  const niftiURL = volumeId.substring(NIFTI_LOADER_SCHEME.length + 1);
+
+  const progress = (loaded, total) => {
+    const data = { volumeId, loaded, total };
+    triggerEvent(eventTarget, Events.NIFTI_VOLUME_PROGRESS, {
+      data,
+    });
+  };
+
+  const onLoad = () => {
+    const data = { volumeId };
+    triggerEvent(eventTarget, Events.NIFTI_VOLUME_LOADED, {
+      data,
+    });
+  };
+
+  const controller = new AbortController();
+  const signal = controller.signal;
+
+  urlsMap.set(niftiURL, { controller, loading: true });
+
+  let niftiBuffer = (await fetchArrayBuffer(
+    niftiURL,
+    progress,
+    signal,
+    onLoad
+  )) as ArrayBuffer;
+
+  urlsMap.delete(niftiURL);
+
+  let niftiHeader = null;
+  let niftiImage = null;
+
+  if (NiftiReader.isCompressed(niftiBuffer)) {
+    niftiBuffer = NiftiReader.decompress(niftiBuffer);
+  }
+
+  if (NiftiReader.isNIFTI(niftiBuffer)) {
+    niftiHeader = NiftiReader.readHeader(niftiBuffer);
+    niftiImage = NiftiReader.readImage(niftiHeader, niftiBuffer);
+  }
+
+  const typedNiftiArray = getTypedNiftiArray(
+    niftiHeader.datatypeCode,
+    niftiImage
+  );
+
+  // Convert to LPS for display in OHIF.
+
+  const { orientation, origin, spacing } = rasToLps(niftiHeader);
+  invertDataPerFrame(niftiHeader.dims.slice(1, 4), typedNiftiArray);
+
+  scaleNiftiArray(typedNiftiArray, niftiHeader);
+
+  const volumeMetadata = makeVolumeMetadata(
+    niftiHeader,
+    orientation,
+    typedNiftiArray
+  );
+
+  const scanAxisNormal = vec3.create();
+  vec3.set(scanAxisNormal, orientation[6], orientation[7], orientation[8]);
+
+  const {
+    BitsAllocated,
+    PixelRepresentation,
+    PhotometricInterpretation,
+    ImageOrientationPatient,
+    Columns,
+    Rows,
+  } = volumeMetadata;
+
+  const rowCosineVec = vec3.fromValues(
+    ImageOrientationPatient[0],
+    ImageOrientationPatient[1],
+    ImageOrientationPatient[2]
+  );
+  const colCosineVec = vec3.fromValues(
+    ImageOrientationPatient[3],
+    ImageOrientationPatient[4],
+    ImageOrientationPatient[5]
+  );
+
+  const { dims } = niftiHeader;
+
+  const numFrames = dims[3];
+
+  // Spacing goes [1] then [0], as [1] is column spacing (x) and [0] is row spacing (y)
+  const dimensions = <Types.Point3>[Columns, Rows, numFrames];
+  const direction = new Float32Array([
+    rowCosineVec[0],
+    rowCosineVec[1],
+    rowCosineVec[2],
+    colCosineVec[0],
+    colCosineVec[1],
+    colCosineVec[2],
+    scanAxisNormal[0],
+    scanAxisNormal[1],
+    scanAxisNormal[2],
+  ]) as Types.Mat3;
+  const signed = PixelRepresentation === 1;
+
+  // Check if it fits in the cache before we allocate data
+  // TODO Improve this when we have support for more types
+  // NOTE: We use 4 bytes per voxel as we are using Float32.
+  let bytesPerVoxel = 1;
```

Comment:
And a fair bit of this could also be put into functions that provide specific bits of data/usage, and can have their own documentation and maybe even their own unit tests.